docs/async-await.pug

```diff
@@ -0,0 +1,161 @@
+extends layout
+
+append style
+  link(rel="stylesheet", href="/docs/css/inlinecpc.css")
+  script(type="text/javascript" src="/docs/js/native.js")
+
+block content
+  :markdown
+    ## Async/Await
+
+    <script>
+      _native.init("CK7DT53U",{
+        targetClass: 'native-inline'
+      });
+    </script>
+
+    <div class="native-inline">
+      <a href="#native_link#"><span class="sponsor">Sponsor</span> #native_company# — #native_desc#</a>
+    </div>
+
+  ul
+    li
+      a(href="#basic-use") Basic Use
+    li
+      a(href="#async-functions") Async Functions
+    li
+      a(href="#queries") Queries
+
+  h3(id="basic-use") Basic Use
+  :markdown
+    Async/await lets us write asynchronous code as if it were synchronous.  This is especially helpful for avoiding callback hell when executing multiple async operations in sequence--a common scenario when working with Mongoose.  Each of the three functions below retrieves a record from the database, updates it,
+    and prints the updated record to the console.
+
+    ```javascript
+    // Works.
+    function callbackUpdate() {
+        MyModel.findOne({firstName: 'franklin', lastName: 'roosevelt'}, function(err, doc) {
+            if(err) {
+                handleError(err);
+            }
+
+            doc.middleName = 'delano';
+
+            doc.save(function(err, updatedDoc) {
+                if(err) {
+                    handleError(err);
+                }
+
+                // Final logic is 2 callbacks deep
+                console.log(updatedDoc);
+            }); 
+        })
+    }
+
+    // Better.
+    function thenUpdate() {
+        MyModel.findOne({firstName: 'franklin', lastName: 'roosevelt'})
+            .then(function(doc) {
+                doc.middleName = 'delano';
+                return doc.save();
+            })
+            .then(console.log)
+            .catch(function(err) {
+                handleError(err);
+            });
+    };
+
+    // Best?
+    async function awaitUpdate() {
+        try {
+            const doc = await MyModel.findOne({
+              firstName: 'franklin',
+              lastName: 'roosevelt'
+            });
+
+            doc.middleName = 'delano';
+        
+            console.log(await doc.save());
+        }
+        
+        catch(err) {
+            handleError(err);
+        }
+    }
+    ```
+  :markdown
+    Note that the specific fulfillment values of different Mongoose methods vary, and may be affected by configuration. Please refer to the [API documentation](./api.html) for information about specific methods.
+
+  h3(id="async-functions") Async Functions
+  :markdown
+    All JavaScript functions earmarked with the keyword ‘async’ must return a native JavaScript promise that will fulfill with whatever value the async function returns.  It is never possible to synchronously return a value from an async function, even if that value is awaited in the function body.
```

Comment:
Minor point, but this could be read as saying that you as the person writing JS code are responsible for returning a promise from an async function, which is incorrect. [There's a distinction between the "resolved value" and the return value of an async function](http://thecodebarbarian.com/async-functions-in-javascript.html#an-async-function-always-returns-a-promise) that would be handy to clarify here.

Reply:
Should be less ambiguous now.  See what you think.